aptos-move/aptos-vm/src/move_vm_ext.rs

```diff
@@ -0,0 +1,168 @@
+///! taken care of after session finish.
+use crate::{
+    access_path_cache::AccessPathCache, aptos_vm_impl::convert_changeset_and_events_cached,
+    natives::aptos_natives, transaction_metadata::TransactionMetadata,
+};
+use aptos_crypto::{hash::CryptoHash, HashValue};
+use aptos_crypto_derive::{BCSCryptoHash, CryptoHasher};
+use aptos_types::{
+    block_metadata::BlockMetadata,
+    transaction::{ChangeSet, SignatureCheckedTransaction},
+};
+use move_binary_format::errors::VMResult;
+use move_core_types::{
+    account_address::AccountAddress,
+    effects::{ChangeSet as MoveChangeSet, Event as MoveEvent},
+    resolver::MoveResolver,
+    vm_status::VMStatus,
+};
+use move_vm_runtime::{
+    move_vm::MoveVM, native_functions::NativeContextExtensions, session::Session,
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    convert::TryInto,
+    ops::{Deref, DerefMut},
+};
+
+pub struct MoveVmExt {
+    inner: MoveVM,
+}
+
+impl MoveVmExt {
+    pub fn new() -> VMResult<Self> {
+        Ok(Self {
+            inner: MoveVM::new(aptos_natives())?,
+        })
+    }
+
+    pub fn new_session<'r, S: MoveResolver>(
+        &self,
+        remote: &'r S,
+        _session_id: SessionId,
+    ) -> SessionExt<'r, '_, S> {
+        // TODO: install table extension
+        let extensions = NativeContextExtensions::default();
+
+        SessionExt {
+            inner: self.inner.new_session_with_extensions(remote, extensions),
+        }
+    }
+}
+
+impl Deref for MoveVmExt {
+    type Target = MoveVM;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+#[derive(BCSCryptoHash, CryptoHasher, Deserialize, Serialize)]
+pub enum SessionId {
+    Txn {
+        sender: AccountAddress,
+        sequence_number: u64,
+    },
+    BlockMeta {
+        // block id
+        id: HashValue,
+    },
+    Genesis {
+        // id to identify this specific genesis build
+        id: HashValue,
+    },
+    // For those runs that are not a transaction and the output of which won't be committed.
+    Void,
+}
+
+impl SessionId {
+    pub fn txn(txn: &SignatureCheckedTransaction) -> Self {
+        Self::Txn {
+            sender: txn.sender(),
+            sequence_number: txn.sequence_number(),
+        }
+    }
+
+    pub fn txn_meta(txn_data: &TransactionMetadata) -> Self {
+        Self::Txn {
+            sender: txn_data.sender,
+            sequence_number: txn_data.sequence_number,
+        }
+    }
+
+    pub fn genesis(id: HashValue) -> Self {
+        Self::Genesis { id }
+    }
+
+    pub fn block_meta(block_meta: &BlockMetadata) -> Self {
+        Self::BlockMeta {
+            id: block_meta.id(),
+        }
+    }
+
+    pub fn void() -> Self {
+        Self::Void
+    }
+
+    pub fn as_uuid(&self) -> u128 {
+        u128::from_be_bytes(self.hash().as_ref()[..16].try_into().unwrap())
+    }
+}
+
+pub struct SessionExt<'r, 'l, S> {
```

Comment:
I think the vm and the session are highly related to each other, for example I view `new_session_with_extensions()` and `finish_with_extensions()` as a pair, and the file is still small. So I preferred them to be together. Lmk if you feel strongly about separating them (because I'm not)